src/kernel/src/fs/mount.rs

```diff
@@ -140,8 +150,94 @@ bitflags! {
     }
 }
 
+pub struct MountOpts(HashMap<&'static str, MountOpt>);
+
+impl MountOpts {
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    pub fn insert(&mut self, k: &'static str, v: impl Into<MountOpt>) {
+        self.0.insert(k, v.into());
+    }
+
+    pub fn remove<T>(&mut self, name: &'static str) -> Option<Result<T, MountOptError>>
+    where
+        T: TryFrom<MountOpt, Error = MountOpt>,
+    {
+        let opt = self.0.remove(name)?;
+
+        let res = opt
+            .try_into()
+            .map_err(|opt| MountOptError::new::<T>(name, opt));
+
+        Some(res)
+    }
+
+    pub fn retain(&mut self, mut f: impl FnMut(&str, &mut MountOpt) -> bool) {
+        self.0.retain(|k, v| f(*k, v));
+    }
+}
+
+#[derive(EnumConversions, Debug)]
+pub enum MountOpt {
+    Bool(bool),
+    I32(i32),
+    U64(u64),
+    Usize(usize),
+    Str(Box<str>),
+    VPath(VPathBuf),
+    Path(PathBuf),
+    Param(Arc<Param>),
+    Gid(Gid),
+    Uid(Uid),
+    Mode(Mode),
+}
+
+impl MountOpt {
+    pub fn as_bool(&self) -> Option<bool> {
+        match self {
+            Self::Bool(v) => Some(*v),
+            _ => None,
+        }
+    }
+}
+
+impl From<&str> for MountOpt {
+    fn from(v: &str) -> Self {
+        Self::Str(v.into())
+    }
+}
+
+impl From<String> for MountOpt {
+    fn from(v: String) -> Self {
+        Self::Str(v.into_boxed_str())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum MountOptError {
+    #[error("mount opt \"{optname}\" is of wrong type: expected {expected}, got: {got:?}")]
+    UnexpectedOptType {
+        optname: &'static str,
+        expected: &'static str,
+        got: MountOpt,
+    },
```

Comment:
If `MountOptError` never going to have more than one variant better to change it to a struct and implement `Error` manually.